### PR TITLE
createMaterialId refactored

### LINKED ISSUE #5364: Polyline material with texture failure
We want to develop a material which have texture uniform for a polyline. But cesium crush with an error message "TypeError: cyclic object value createMaterialId@........". Texture objects have cyclic value and createMaterialId function in `PolylineCollection` computes id as below;

```js
function createMaterialId(material) {
        var uniforms = Material._uniformList[material.type];
        var length = uniforms.length;
        scratchUniformArray.length = 2.0 * length;

        var index = 0;
        for (var i = 0; i < length; ++i) {
            var uniform = uniforms[i];
            scratchUniformArray[index] = uniform;
            scratchUniformArray[index + 1] = material._uniforms[uniform]();
            index += 2;
        }

        return material.type + ':' + JSON.stringify(scratchUniformArray);
    }
```

If createMaterialId will be updated as:

```js
function createMaterialId(material) {
        var uniforms = Material._uniformList[material.type];
        var length = uniforms.length;
        scratchUniformArray.length = 2.0 * length;

        var index = 0;
        var uniformValue;
        for (var i = 0; i < length; ++i) {
            var uniform = uniforms[i];
            scratchUniformArray[index] = uniform;
            uniformValue=material._uniforms[uniform]();
            
            if(uniformValue instanceof Texture){
                scratchUniformArray[index + 1] = computeTextureId(uniformValue);
            } else {
                scratchUniformArray[index + 1] = uniformValue;
            }
            index += 2;
        }

        return material.type + ':' + JSON.stringify(scratchUniformArray);
    }
```

problem can be fixed

### PR DESCRIPTION
**Pull request by ateshuseyin**
fixes #5364

### CODE REVIEW

**Comment by pjcozzi** [comment]
Thanks @ateshuseyin!

@bagnell do you want to look at this?

**Comment by bagnell** [comment]
`createTextureId` doesn't uniquely identify the texture. You could have two separate textures where those properties are the exact same.

I think a better way to do this would be to change `Texture` to have a unique id (either a GUID or a number that increases every time a texture is created). Then, add a replacer function to `JSON.stringify` that replaces the texture with the id. See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify

**Comment by bagnell** [comment]
Also, can you add a test and update `CHANGES.md`?

File: Source/Renderer/Texture.js [hunk 1]
```diff
@@ -196,6 +198,7 @@ define([
             sizeInBytes = PixelFormat.textureSizeInBytes(pixelFormat, pixelDatatype, width, height);
         }
 
+        this._id = 'Texture_' + currentId++;
```
Comment (bagnell) [inline Source/Renderer/Texture.js]: Just `currentId++` is fine.

File: Source/Renderer/Texture.js [hunk 2]
```diff
@@ -308,6 +311,12 @@ define([
     };
 
     defineProperties(Texture.prototype, {
+
+        id : {
```
Comment (bagnell) [inline Source/Renderer/Texture.js]: Can you add documentation here? Please also mark `@private`.

File: Source/Scene/PolylineCollection.js [hunk 3]
```diff
@@ -1013,6 +1013,14 @@ define([
         }
     }
 
+    function replacer(key, value) {
+        if(defined(value) && value.hasOwnProperty('id')) {
```
Comment (bagnell) [inline Source/Scene/PolylineCollection.js]: Does `value instanceof Texture` not work here? We should only do this for textures.

**Comment by bagnell** [comment]
Can you add a test that shows issue #5364 was fixed? It should fail in master and pass in this branch. Also update `CHANGES.md`.

File: Source/Scene/PolylineCollection.js (hunk 3 above)
Comment (ateshuseyin) [inline Source/Scene/PolylineCollection.js]: It works. I thougt that if there is a id, it should be unique. It is uncertain.

**Comment by ateshuseyin** [comment]
Deal. I didn't post a comment because not ready yet. I will post when ready

**Comment by ateshuseyin** [comment]
Ready.

**Comment by pjcozzi** [comment]
Thanks again @ateshuseyin!

@bagnell can you take another look and merge when ready?

**Comment by bagnell** [comment]
@ateshuseyin I requested some changes in the comments. Do you have time to update the code?

**Comment by ateshuseyin** [comment]
I updated my code. @bagnell could you help me to write test.

**Comment by hpinkos** [comment]
@bagnell I pushed some changes here.  Does this PR need anything else or is it good to go?

**Comment by bagnell** [comment]
Looks good. Thanks @ateshuseyin and @hpinkos!

**Pull request merged**